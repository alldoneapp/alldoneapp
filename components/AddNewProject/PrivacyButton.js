import React, { useState } from 'react'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { View, Text, StyleSheet } from 'react-native'
import Icon from '../Icon'
import { colors } from '../styles/global'
import Popover from 'react-tiny-popover'
import ProjectPrivacyModal from '../ProjectDetailedView/ProjectProperties/PrivacyProperty/ProjectPrivacyModal'
import ProjectHelper from '../SettingsView/ProjectsSettings/ProjectHelper'
import store from '../../redux/store'
import { hideFloatPopup, showFloatPopup } from '../../redux/actions'

export default function PrivacyButton({ value, setPrivacy, disabled }) {
    const [isOpen, setIsOpen] = useState(false)

    const hideModal = () => {
        setTimeout(() => {
            setIsOpen(false)
            store.dispatch(hideFloatPopup())
        })
    }

    const showModal = () => {
        setIsOpen(true)
        store.dispatch(showFloatPopup())
    }

    return (
        <Popover
            isOpen={isOpen}
            onClickOutside={() => setIsOpen(false)}
            align={'center'}
            position={['right', 'bottom', 'top']}
            content={
                <ProjectPrivacyModal
                    setIsOpen={hideModal}
                    project={{ isShared: value }}
                    setSharedProperty={setPrivacy}
                />
            }
        >
            <TouchableOpacity
                style={localStyles.container}
                onPress={isOpen ? hideModal : showModal}
                disabled={disabled}
            >
                <View style={localStyles.button}>
                    <Icon name={ProjectHelper.getProjectPrivacyIcon(value)} size={20} color={colors.Grey100} />
                </View>
            </TouchableOpacity>
        </Popover>
    )
}

const localStyles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        borderRadius: 4,
        backgroundColor: colors.Secondary200,
    },
    button: {
        width: 40,
        height: 40,
        paddingHorizontal: 10,
        flexDirection: 'row',
        alignItems: 'center',
    },
})
