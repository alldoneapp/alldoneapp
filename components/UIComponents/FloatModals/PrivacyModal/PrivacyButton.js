import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import Popover from 'react-tiny-popover'
import Hotkeys from 'react-hot-keys'

import PrivacyModal from './PrivacyModal'
import { hideFloatPopup, showFloatPopup } from '../../../../redux/actions'
import { execShortcutFn } from '../../../../utils/HelperFunctions'
import GhostButton from '../../../UIControls/GhostButton'
import ButtonUsersGroup from './ButtonUsersGroup'
import { FEED_PUBLIC_FOR_ALL, FEED_USER_OBJECT_TYPE } from '../../../Feeds/Utils/FeedsConstants'
import ProjectHelper from '../../../SettingsView/ProjectsSettings/ProjectHelper'
import ContactsHelper from '../../../ContactsView/Utils/ContactsHelper'
import { translate } from '../../../../i18n/TranslationService'

export default function PrivacyButton({
    projectId,
    object,
    disabled,
    style,
    objectType,
    savePrivacyBeforeSaveObject,
    shortcutText,
    inEditComponent,
    callback,
    onDismissPopup,
}) {
    const dispatch = useDispatch()
    const smallScreen = useSelector(state => state.smallScreen)
    const blockShortcuts = useSelector(state => state.blockShortcuts)
    const [visiblePopover, setVisiblePopover] = useState(false)

    const hidePopover = () => {
        setVisiblePopover(false)
        dispatch(hideFloatPopup())
        if (onDismissPopup) onDismissPopup()
    }

    const delayHidePopover = e => {
        if (e) {
            e.preventDefault()
            e.stopPropagation()
        }
        setTimeout(async () => {
            hidePopover()
        }, 200)
    }

    const showPopover = () => {
        setVisiblePopover(true)
        dispatch(showFloatPopup())
    }

    if (objectType === FEED_USER_OBJECT_TYPE) {
        const projectIndex = ProjectHelper.getProjectIndexById(projectId)
        ContactsHelper.getAndAssignUserPrivacy(projectIndex, object)
    }

    const { isPublicFor } = object

    return (
        <Popover
            content={
                <PrivacyModal
                    object={object}
                    objectType={objectType}
                    projectId={projectId}
                    closePopover={hidePopover}
                    delayClosePopover={delayHidePopover}
                    savePrivacyBeforeSaveObject={savePrivacyBeforeSaveObject}
                    callback={callback}
                />
            }
            onClickOutside={delayHidePopover}
            isOpen={visiblePopover}
            position={['left', 'right', 'top', 'bottom']}
            padding={4}
            align={'end'}
            contentLocation={smallScreen ? null : undefined}
        >
            <Hotkeys
                keyName={`alt+${shortcutText}`}
                disabled={disabled || blockShortcuts}
                onKeyDown={(sht, event) => execShortcutFn(this.buttonRef, showPopover, event)}
                filter={e => true}
            >
                <GhostButton
                    ref={ref => (this.buttonRef = ref)}
                    title={
                        inEditComponent && smallScreen
                            ? null
                            : isPublicFor.includes(FEED_PUBLIC_FOR_ALL)
                            ? translate('Public')
                            : translate('Private')
                    }
                    type={'ghost'}
                    noBorder={inEditComponent && smallScreen}
                    icon={
                        isPublicFor.includes(FEED_PUBLIC_FOR_ALL) ? (
                            'unlock'
                        ) : (
                            <ButtonUsersGroup projectId={projectId} users={isPublicFor} />
                        )
                    }
                    buttonStyle={style}
                    onPress={showPopover}
                    disabled={disabled}
                    shortcutText={shortcutText}
                    accessible={false}
                />
            </Hotkeys>
        </Popover>
    )
}
