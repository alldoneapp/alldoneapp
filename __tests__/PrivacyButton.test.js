import React from 'react'
import PrivacyButton from '../components/AddNewProject/PrivacyButton'
import store from '../redux/store'
import { Provider } from 'react-redux'
import { Platform } from 'react-native'

import renderer from 'react-test-renderer'

jest.mock('firebase', () => ({ firestore: {} }))

describe('PrivacyButton component', () => {
    beforeEach(() => {
        jest.useFakeTimers()
    })
    describe('PrivacyButton empty snapshot test', () => {
        it('should render correctly', () => {
            const tree = renderer
                .create(
                    <Provider store={store}>
                        <PrivacyButton />
                    </Provider>
                )
                .toJSON()
            expect(tree).toMatchSnapshot()
        })

        it('should render correctly when is web', () => {
            Platform.OS = 'web'
            const tree = renderer
                .create(
                    <Provider store={store}>
                        <PrivacyButton />
                    </Provider>
                )
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })
})
