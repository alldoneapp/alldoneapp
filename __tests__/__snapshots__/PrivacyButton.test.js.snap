// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivacyButton component PrivacyButton empty snapshot test should render correctly 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1A3289",
        "borderRadius": 4,
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 16,
        "paddingRight": 24,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#FAFBFB",
              "fontFamily": "alldone",
              "fontSize": 20,
              "opacity": 1,
            }
          }
        >
          
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Roboto-Medium",
              "fontSize": 14,
              "fontStyle": "normal",
              "letterSpacing": 0.8,
              "lineHeight": 14,
            },
            Object {
              "color": "#FAFBFB",
              "paddingLeft": 14,
            },
          ]
        }
      >
        Public
      </Text>
    </View>
  </View>
</RNGestureHandlerButton>
`;

exports[`PrivacyButton component PrivacyButton empty snapshot test should render correctly when is web 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1A3289",
        "borderRadius": 4,
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 16,
        "paddingRight": 24,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#FAFBFB",
              "fontFamily": "alldone",
              "fontSize": 20,
              "opacity": 1,
            }
          }
        >
          
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Roboto-Medium",
              "fontSize": 14,
              "fontStyle": "normal",
              "letterSpacing": 0.8,
              "lineHeight": 14,
            },
            Object {
              "color": "#FAFBFB",
              "paddingLeft": 14,
            },
          ]
        }
      >
        Public
      </Text>
    </View>
  </View>
</RNGestureHandlerButton>
`;
